File: src/main/java/com/pedigree/util/DirtyObservableList.java
package com.pedigree.util;

import java.util.ArrayList;
import java.util.Collection;
import java.util.function.UnaryOperator;

public class DirtyObservableList<E> extends ArrayList<E> {

    public DirtyObservableList() {
        super();
    }

    public DirtyObservableList(Collection<? extends E> c) {
        super(c);
    }

    private void dirty() {
        DirtyFlag.setModified();
    }

    @Override
    public boolean add(E e) {
        boolean r = super.add(e);
        if (r) dirty();
        return r;
    }

    @Override
    public void add(int index, E element) {
        super.add(index, element);
        dirty();
    }

    @Override
    public boolean addAll(Collection<? extends E> c) {
        boolean r = super.addAll(c);
        if (r) dirty();
        return r;
    }

    @Override
    public boolean addAll(int index, Collection<? extends E> c) {
        boolean r = super.addAll(index, c);
        if (r) dirty();
        return r;
    }

    @Override
    public E remove(int index) {
        E e = super.remove(index);
        dirty();
        return e;
    }

    @Override
    public boolean remove(Object o) {
        boolean r = super.remove(o);
        if (r) dirty();
        return r;
    }

    @Override
    public boolean removeAll(Collection<?> c) {
        boolean r = super.removeAll(c);
        if (r) dirty();
        return r;
    }

    @Override
    public boolean retainAll(Collection<?> c) {
        boolean r = super.retainAll(c);
        if (r) dirty();
        return r;
    }

    @Override
    public void clear() {
        if (!isEmpty()) {
            super.clear();
            dirty();
        } else {
            super.clear();
        }
    }

    @Override
    public E set(int index, E element) {
        E prev = super.set(index, element);
        // consider any set as modification
        dirty();
        return prev;
    }

    @Override
    public void replaceAll(UnaryOperator<E> operator) {
        super.replaceAll(operator);
        dirty();
    }
}

File: src/main/kotlin/com/family/tree/util/DirtyObservableList.kt
package com.family.tree.util

import java.util.ArrayList

/**
 * ArrayList that marks the global DirtyFlag on any mutating operation.
 * Mirrors the original Java API for seamless interop.
 */
class DirtyObservableList<E> : ArrayList<E> {

    constructor() : super()
    constructor(c: Collection<out E>) : super(c)

    private fun dirty() = DirtyFlag.setModified()

    override fun add(element: E): Boolean {
        val r = super.add(element)
        if (r) dirty()
        return r
    }

    override fun add(index: Int, element: E) {
        super.add(index, element)
        dirty()
    }

    override fun addAll(elements: Collection<E>): Boolean {
        val r = super.addAll(elements)
        if (r) dirty()
        return r
    }

    override fun addAll(index: Int, elements: Collection<E>): Boolean {
        val r = super.addAll(index, elements)
        if (r) dirty()
        return r
    }

    override fun removeAt(index: Int): E {
        val e = super.removeAt(index)
        dirty()
        return e
    }

    override fun remove(element: E): Boolean {
        val r = super.remove(element)
        if (r) dirty()
        return r
    }

    override fun removeAll(elements: Collection<E>): Boolean {
        val r = super.removeAll(elements)
        if (r) dirty()
        return r
    }

    override fun retainAll(elements: Collection<E>): Boolean {
        val r = super.retainAll(elements)
        if (r) dirty()
        return r
    }

    override fun clear() {
        if (isNotEmpty()) {
            super.clear()
            dirty()
        } else {
            super.clear()
        }
    }

    override fun set(index: Int, element: E): E {
        val prev = super.set(index, element)
        dirty()
        return prev
    }

    override fun replaceAll(operator: java.util.function.UnaryOperator<E>) {
        super.replaceAll(operator)
        dirty()
    }
}
